sweep/aggregator.go

```diff
@@ -742,6 +747,64 @@ func (b *BudgetAggregator) sortInputs(inputs []SweeperInput) []SweeperInput {
 	return sortedInputs
 }
 
+// splitOnLocktime splits the list of inputs based on their locktime.
+//
+// TODO(yy): this is a temporary hack as the blocks are not synced among the
+// contractcourt and the sweeper.
+func splitOnLocktime(inputs []SweeperInput) map[uint32][]SweeperInput {
+	result := make(map[uint32][]SweeperInput)
+	noLocktimeInputs := make([]SweeperInput, 0, len(inputs))
+
+	// mergeLocktime is the locktime that we use to merge all the
+	// nolocktime inputs into.
+	var mergeLocktime uint32
+
+	// Iterate all inputs and split them based on their locktimes.
+	for _, inp := range inputs {
+		locktime, required := inp.RequiredLockTime()
+		if !required {
+			log.Tracef("No locktime required for input=%v",
+				inp.OutPoint())
+
+			noLocktimeInputs = append(noLocktimeInputs, inp)
+			continue
+		}
+
+		log.Tracef("Split input=%v on locktime=%v", inp.OutPoint(),
+			locktime)
+
+		// Init the slice if not found.
+		inputList, ok := result[locktime]
+		if !ok {
+			inputList = make([]SweeperInput, 0)
+		}
+
+		// Add the input to the list.
+		inputList = append(inputList, inp)
+
+		// Update the map.
+		result[locktime] = inputList
+
+		// Update the merge locktime.
+		mergeLocktime = locktime
```

Comment:
Am I to understand that this means that we favor grouping unspecified locktime inputs with the last input in this list? IIRC that means that we group these inputs in with the one with the highest feerate and that doesn't make sense to me as a choice. It's a bit hard to understand the intent behind this algorithm so it's possible I misunderstood it but I believe we choose the last locktime supplied. Is that right? If so, does this choice make sense?

Reply:
As explained above, the grouping logic is to make sure different locktimes don't get mixed. On the other hand, the locktime chosen here doesn't correlate to the fee rate. Usually the larger the locktime, the more likely you'd use a smaller starting fee rate, but we don't choose the merge locktime here based on values, but based on the existence. And yea, don't think this grouping logic will be hit in practice - but just in case.

Reply:
Ah I understand now. We always merge the no locktime inputs into the **last** locktime we found in this list traversal.